omigami/feature_selector.py

```diff
@@ -22,68 +37,198 @@ def __init__(
+            groups = np.arange(size)
+        return groups
 
-    def _make_feature_evaluator(self, input_data: InputData) -> FeatureEvaluator:
-        return FeatureEvaluator(
-            input_data,
-            self.n_outer,
-            self.n_inner,
-            self.estimator,
-            self.metric,
-            self.random_state,
+    def _deferred_run_outer_loop(
+        self,
+        input_data: InputDataset,
+        outer_split: Split,
+        data_splitter: DataSplitter,
+        executor: Executor,
+    ) -> Union[Future, OuterLoopResults]:
+        if executor is None:
+            return self._run_outer_loop(input_data, outer_split, data_splitter)
+        return executor.submit(
+            self._run_outer_loop, input_data, outer_split, data_splitter
+        )
+
+    def _run_outer_loop(
+        self, input_data: InputDataset, outer_split: Split, data_splitter: DataSplitter,
+    ) -> OuterLoopResults:
+
+        raw_feature_elim_results = {}
+        feature_set = list(range(input_data.n_features))
+
+        while len(feature_set) >= self._minimum_features:
+            inner_results = []
+
+            for inner_split in data_splitter.iter_inner_splits(outer_split):
+                inner_loop_data = data_splitter.split_data(
+                    input_data, inner_split, feature_set
+                )
+
+                feature_evaluation_results = self.feature_evaluator.evaluate_features(
+                    inner_loop_data, feature_set
+                )
+
+                inner_results.append(feature_evaluation_results)
+
+            raw_feature_elim_results[tuple(feature_set)] = inner_results
+            feature_set = self._remove_features(feature_set, inner_results)
+
+        feature_elimination_results = self.post_processor.process_feature_elim_results(
+            raw_feature_elim_results
+        )
+        outer_loop_results = self.create_outer_loop_results(
+            feature_elimination_results, input_data, outer_split, data_splitter
         )
 
-    def _make_outer_loop(self, input_data: InputData) -> OuterLoop:
-        feature_evaluator = self._make_feature_evaluator(input_data)
-        return OuterLoop(
-            self.n_outer,
-            feature_evaluator,
-            self.features_dropout_rate,
-            self.robust_minimum,
+        return outer_loop_results
+
+    def _remove_features(
+        self, features: List[int], results: InnerLoopResults
+    ) -> List[int]:
+        features_to_keep = int(np.floor(len(features) * self.keep_fraction))
+        features = self._select_n_best(results, features_to_keep)
+        return features
+
+    @staticmethod
+    def _select_n_best(inner_loop_result: InnerLoopResults, keep_n: int) -> List[int]:
+        if keep_n < 1:
+            return []
+        ranks = [r.ranks for r in inner_loop_result]
+        avg_ranks = average_ranks(ranks)
+        return get_best_n_features(avg_ranks, keep_n)
+
+    def create_outer_loop_results(
```

Comment:
I think if we construct OuterLoopResults in `_run_outer_loop` method we can get rid of this method here.

Reply:
I think it is good now after the other changes